### PR TITLE
add pruning method: Filter Pruning via Geometric Median for Deep Convolutional Neural Networks Acceleration

### PR DESCRIPTION
**Pull request by ArmandXiao**
add `class FPGMStructured`
add `function FPGM_structured()`
add `function _validate_distance_type()`
add `function _compute_distance()`

Implement method mentioned in issue #39765

---
FPGMSparsifier Implement with the new pytorch pruning API torch.ao.pruning.
It is a structured pruning method, and it is added under torch.ao.pruning._experimental. Test cases are added at `test_structured_sparsifier.py`.

### CODE REVIEW

**Comment by ArmandXiao** [comment]
> Looks like this PR hasn't been updated in a while so we're going to go ahead and mark this as `Stale`. Feel free to remove the `Stale` label if you feel this was a mistake. If you are unable to remove the `Stale` label please contact a maintainer in order to do so. If you want the bot to never mark this PR stale again, add the `no-stale` label.`Stale` pull requests will automatically be closed after 30 days of inactivity.

Dear @H-Huang , can you help to add a `non-stale` tag for this PR?
Many thanks in advance.

**Comment by ArmandXiao** [comment]
> torch.ao.pruning

Noted! Thanks for the information :)

**Comment by ArmandXiao** [comment]
Hi @jcaip, please help to check the code. Many thanks!

File: test/ao/sparsity/test_structured_sparsifier.py [hunk 1]
```diff
@@ -916,3 +917,421 @@ def test_prune_lstm_layernorm_linear_single_layer(self):
         # also check that output of linear is the same shape, this means we've resized
         # linear columns correctly.
         assert out_expected.shape == out_pruned.shape
+
+class TestFPGMSparsifier(TestCase):
```
Comment (jcaip) [inline test/ao/sparsity/test_structured_sparsifier.py:921]: It's not necessary to have all these specific pattern checks, as they are already being checked in test_sparse_semi_structured (for the base class). I would remove them, as it'll just add extra time for CI and not provide much signal. 

At the same time I don't see an explicit test for the implementation of FPGM? I'd like to see a test that makes sure `update_mask` is implemented correctly, like [here](https://github.com/pytorch/pytorch/blob/33b855e9069ddc98399c0dfd362f18df9503b66f/test/ao/sparsity/test_structured_sparsifier.py#L75). This should be the only test you need.

File: torch/ao/pruning/_experimental/pruner/fpgm_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+class FPGMSparsifier(BaseStructuredSparsifier):
+    r"""Filter Pruning via Geometric Median (FPGM) Structured Sparsifier
+    This sparsifier prune fliter (row) in a tensor according to distances among filters according to
+    "Filter Pruning via Geometric Median for Deep Convolutional Neural Networks Acceleration"
+    https://arxiv.org/abs/1811.00250.
+
+    This sparsifier is controlled by three variables:
+    1. `sparsity_level` defines the number of filters (rows) that are zeroed-out.
+    2. `dim` defines the dimension of the tensor to prune. Default: 0 (filter pruning).
```
Comment (jcaip) [inline torch/ao/pruning/_experimental/pruner/fpgm_sparsifier.py:18]: I think the paper only supports Filter pruning, which is also true of our mask parameterization implementation. Does the paper also mention kernel / subkernel pruning as well? If not let's remove this extra parameter and the code used to handle it

File: test/ao/sparsity/test_structured_sparsifier.py (hunk 1 above)
Comment (jcaip) [reply test/ao/sparsity/test_structured_sparsifier.py:921]: A test for _compute_distances would also be nice

**Review by jcaip — CHANGES_REQUESTED** [review]
@ArmandXiao Thanks for working on this!

This looks pretty good, but I'd like to see a test case that explicitly tests the `update_mask` implementation, instead of testing that the sparsifier is being applied to the patterns correctly. 

Additionally, from my understanding the paper only implements filter pruning, which is also the only supported pruning method by our parameterization. If this is the case let's remove some of the custom dimension handling logic, since it will always be 0.

**Comment by ArmandXiao** [comment]
Hi @jcaip , thanks for your review.

I have made the following changes:
In the [second commit: Remove unnecessary parts, and small fixes](https://github.com/pytorch/pytorch/commit/d3b5dad4846cecccb890ae70d07c23aeaa439923)
1. For `test_structured_sparsifier.py`:
remove unnecessary pattern checks

2. For `FPGM_sparsifier.py`:
remove `dim` parameter and associated comments

3. For `FPGM_sparsifier.py`:
replace conditional statement with clamp operation

In the [third commit: Add test for compute_distance(), update_mask()](https://github.com/pytorch/pytorch/commit/d0605194490c2522879dc21987263b83b171397f):
1. For `test_structured_sparsifier.py`:
add test for `compute_distance()`

2. For `test_structured_sparsifier.py`:
add test for `update_mask()`

In the [last commit: rename FPGMSparsifier to FPGMPruner](https://github.com/pytorch/pytorch/commit/29c630fa060d3e0c9464821673b1d810ca0f61f0), I renamed `FPGMSparsifier` to `FPGMPruner` in accordance to other implementations.

---

In addition,  I kept receiving linux/android/ios build errors.
Do you have any clue about this?

Your time is appreciated! Many thanks in advance.

**Comment by jcaip** [comment]
@ArmandXiao Thanks for the changes

FYI you can see the results for test runs on the [HUD](https://hud.pytorch.org/pytorch/pytorch/pull/95689?sha=29c630fa060d3e0c9464821673b1d810ca0f61f0). 

I don't see any of the iOS/android errors you mentioned? It looks like the test you wrote were failing on some import errors. You can ignore some errors like dynamo / distributed errors because those are trunk (main) being broken or flaky tests.

**Comment by ArmandXiao** [comment]
Hi @jcaip , I have added and passed the local test in the latest commit.
Can you help to approve the auto-check from the Pytorch bot?

Many thanks in advance.

**Comment by jcaip** [comment]
@ArmandXiao Thanks - looks like just one more test failing on a device mismatch! Let me know when you fix that and I'll rerun + approve.

File: test/ao/sparsity/test_structured_sparsifier.py [hunk 3]
```diff
@@ -916,3 +917,123 @@ def test_prune_lstm_layernorm_linear_single_layer(self):
...
+        assert torch.isclose(pruned_model.conv2d2.weight, expected_conv2, rtol=1e-05, atol=1e-07).all()
+
+    def test_update_mask(self):
+        weights = torch.tensor([3.0, 0.1])
+        expected_conv1 = torch.ones((2, 1, 3, 3)) * weights[:, None, None, None]
+
+        weights = torch.tensor([7.0, 0.4])
+        expected_conv2 = torch.ones((2, 2, 3, 3)) * weights[:, None, None, None]
+
+        for device in DEVICES:
```
Comment (jcaip) [inline test/ao/sparsity/test_structured_sparsifier.py:1037]: @ArmandXiao I think you just need to move expected_conv1 and expected_conv2 to the right device.

Comment (ArmandXiao) [reply test/ao/sparsity/test_structured_sparsifier.py:1040]: Hi @jcaip , thank you a lot for the information.

In the [latest commit](https://github.com/pytorch/pytorch/pull/95689/commits/1c98cbece24122e3a13c44397f7586aa81ec2e6a), I have move `expected_conv1` and `expected_conv2` to the correct device before the test cases.

**Review by jcaip — APPROVED** [review]

**Comment by ArmandXiao** [comment]
Hi @jcaip , last pytorch check failed at SSH setup, see [HUD](https://hud.pytorch.org/pytorch/pytorch/pull/95689?sha=3a940c3f0ae7fc9e58b3610f3ddb1240b89c9002).

Do you have any clue for this error?

**Comment by jcaip** [comment]
@ArmandXiao Looks all good to me, it might have just been a flaky test.
